---
title: "GSI correlation heatmap developmental modules"
author: "Fabio Sacher"
date: "28.02.2023"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
  html_notebook:
    fig_height: 7
    fig_width: 8
editor_options:
  chunk_output_type: inline
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(
  comment = '', fig.width = 10, fig.height = 7
)
```

```{r libraries, warning = FALSE, message = FALSE, echo = T, results = 'hide'}
library(Seurat)
library(stringr)
library(tibble)
library(dplyr)
library(ggplot2)
library(pheatmap)
library(WGCNA)
library(gridExtra)
source("~/spinal_cord_paper/scripts/heatmap4.R")
```

```{r data-sets}
se_path <- c("Gg_D05_ctrl_seurat_070323",
             "Gg_D07_ctrl_seurat_070323",
             "Gg_ctrl_1_seurat_070323")

clust_col <- read.csv("~/spinal_cord_paper/annotations/broad_cluster_marker_colors.csv") %>% 
  rename(broad = broad_cluster) %>% 
  select(-marker)

```

# Seurat objects

We calulate pseudobulk matrices and normalize them with the gene specificity index, i.e. the average expression of a gene within a cluster (the pseudobulk value) divided by the average expression across all

```{r seurat-objects-and-annotations}
se_obj <- list()
col_table <- list()

for (i in seq(se_path)) {
  # load the data sets
  my.se <- readRDS(paste0("~/spinal_cord_paper/data/", se_path[i], ".rds"))
  annot <- read.csv(list.files("~/spinal_cord_paper/annotations",
                               pattern = str_remove(se_path[i], "_seurat_\\d{6}"),
                               full.names = TRUE))
  
  if(length(table(annot$number)) != length(table(my.se$seurat_clusters))) {
     stop("Number of clusters must be identical!")
  }
  
  # rename for left join
  annot <- annot %>% 
    mutate(fine = paste(fine, number, sep = "_")) %>% 
    mutate(number = factor(number, levels = 1:nrow(annot))) %>% 
    rename(seurat_clusters = number) 
  
  # create index for color coding
  col_table[[i]] <- annot %>%
    left_join(clust_col, by = "broad") %>% 
    select(c("fine", "color"))
  
  # add cluster annotation to meta data
  my.se@meta.data <- my.se@meta.data %>% 
    rownames_to_column("rowname") %>% 
    left_join(annot, by = "seurat_clusters") %>% 
    mutate(fine = factor(fine, levels = annot$fine)) %>% 
    column_to_rownames("rowname")
  
  se_obj[[i]] <- my.se

}

names(se_obj) <- c("Gg_D05_ctrl", "Gg_D07_ctrl", "Gg_D10_ctrl")
names(col_table) <- c("Gg_D05_ctrl", "Gg_D07_ctrl", "Gg_D10_ctrl")

rm(my.se, annot)
```

```{r}
# broad cluster color table
tmp <- do.call(rbind, col_table) %>% 
  rownames_to_column("sample") %>% 
  mutate(sample = substr(sample, 1, 11)) %>% 
  mutate(sample_fine = paste(sample, fine, sep = "_"))
```


# scWGCNA data

Load the devel scWGCNA data.

```{r scWGCNA-data}
# The reference WGCNA data.
WGCNA_data <- readRDS("~/spinal_cord_paper/output/Gg_devel_int_scWGCNA_250723.rds")

# the name of each sample, as they appear in my.files and in the metadata of the combined object
my.samplenames <- c("Gg_D05_ctrl", "Gg_D07_ctrl", "Gg_D10_ctrl")

```

Since the modules are based on the integrated data of all three sets module genes are expressed in all data sets.

# Module gene correlation

Here, we do a correlation matrix / heatmap, to see which cell clusters group togheter. 

```{r correlation, echo=TRUE, fig.height=10, fig.width=12, error=TRUE}

# take the normalized data, of each single-cell object.
my.datExpr <- list()

# Go trough each seurat object
for (i in 1:length(se_obj)) {
 my.datExpr[[i]] <- data.frame(se_obj[[i]]@assays$RNA@data)
}

# Make a new list
my.MEs <- list()
  
# Populate with data frames from each seurat object
for (i in 1:length(se_obj)) {
    
  # Data frame of average module expression, per cell.
  my.MEs[[i]] <- moduleEigengenes(t(my.datExpr[[i]][names(WGCNA_data$dynamicCols),]),
                                 colors = WGCNA_data$dynamicCols)
}


my.modavg <- list()
  
for (i in 1:length(se_obj)) {
    
  #Make the mean per cell clusters
  my.modavg[[i]] <- aggregate(
    my.MEs[[i]]$averageExpr,
    list(se_obj[[i]]@meta.data$seurat_clusters),
    mean
    )
  # Give the cell clusters meaningful names
  my.modavg[[i]]$Group.1 = paste0(
    my.samplenames[i],
    "_",
    levels(se_obj[[i]]@meta.data$fine)
    )
}

#Gather data to plot

#bind the expression data into one dataframe
my.wgmat <- data.table::rbindlist(my.modavg)
#Get rid of the extra column
my.wgmat <- my.wgmat %>% column_to_rownames("Group.1")

```

```{r meta-data}
#Get a dataframe with annotations for all the samples and colors we need.
my.metam <- list()

for (i in seq(se_obj)) {
  my.metam[[i]] <- se_obj[[i]][[]] # meta data
  rownames(my.metam[[i]]) <- paste0(rownames(my.metam[[i]]), "_", i)
}
# all meta data tables concatenated 
my.metam <- do.call(rbind, my.metam)

my.metam$orig.ident <- str_replace_all(my.metam$orig.ident, pattern =  "ctrl_1", "D10_ctrl")

my.metam$sample_celltype <- paste0(my.metam$orig.ident, "_", my.metam$fine)

my.metam <- my.metam %>% 
  rownames_to_column("cell_ID") %>%
  left_join(clust_col, by = "broad") %>%
  mutate(sample_celltype = factor(sample_celltype, levels = rownames(my.wgmat))) %>% 
  column_to_rownames("cell_ID")

#We only need the colors, cluster name and sample name
my.metacols <- my.metam[,c("color", "sample_celltype","orig.ident")]
rm(my.metam)

# get sample colors
my.colsm = c("grey", "grey30", "black")
names(my.colsm) <- c("Gg_D05_ctrl", "Gg_D07_ctrl", "Gg_D10_ctrl")
 
#The colors for the samples and clusters. First the closest to the heatmap. Add a white space to easy the eye and make less confussing
my.heatcols <- as.matrix(data.frame(
  cluster = as.character(tmp$color[match(rownames(my.wgmat),
                                                 tmp$sample_fine)]),
  "." = "white",
  sample = as.character(my.colsm[match(tmp$sample,names(my.colsm))]))
    )


```

## annotations

```{r annot-list}

# names and colors for the heatmap annotation
annot_name <- data.frame(
  "Celltypes" = tmp$sample_fine,
  "Sample"    = tmp$sample,
  row.names = tmp$sample_fine)
  

col_table <- do.call(rbind, col_table) %>% 
  rownames_to_column("sample") %>% 
  mutate(sample = substr(sample, 1,11)) %>% 
  mutate(fine = paste(sample, fine, sep = "_"))

# match color table with annotation
col_table <- col_table[match(annot_name$Celltypes, col_table$fine),]

annot_col <- list(
  Celltypes = col_table$color,
  Sample = c(Gg_D05_ctrl = "#A4A4A4",
             Gg_D07_ctrl = "#515151",
             Gg_D10_ctrl = "#000000")
  )

names(annot_col[[1]]) <- annot_name$Celltypes

```

```{r}
heat_col <- colorRampPalette(colors = c("dodgerblue","dodgerblue4", "white", "red", "darkred"))

#Calculate the distance 1-cor, and then calculate dendograms to cluster the clusters
my.hclust = hclust(as.dist(1-cor(t(my.wgmat), method = "spearman")))

toplot <- cor(t(my.wgmat), method = "spearman")

# lower limit to scale color bar
low_limit <- 100 - abs(round(range(toplot)[1]*100))

corr_heatmap <- pheatmap(toplot,revC = T,
                 main = "spearman correlation of average module eigengene expression by cluster\n Dendrogram based on 'distance 1-cor'",
                 fontsize = 8,  
                 color = heat_col(200)[low_limit:200],
                 show_colnames = F,
                 cluster_rows = my.hclust,
                 cluster_cols = my.hclust,
                 treeheight_row = 0, 
                 annotation_col = annot_name,
                 annotation_colors = annot_col,
                 annotation_legend = F,
                 border_color = NA
                 )
```


```{r plot, fig.width=20, fig.height=20}

#Calculate the distance 1-cor, and then calculate dendograms to cluster the clusters
my.hclust = as.dendrogram(hclust(as.dist(1-cor(t(my.wgmat), method = "spearman"))))

colbar <- c(min(cor(t(my.wgmat), method = "spearman")),
            max(cor(t(my.wgmat), method = "spearman")))

pdf("~/spinal_cord_paper/figures/heatmap_module_gene_devel_spearman.pdf", width = 10, height = 10)
heatmap.4(cor(t(my.wgmat), method = "spearman"),
          col = colorRampPalette(c("gray95","gray5"))(n = 1000),
          trace="none",
          lhei=c(0.2,1.1),
          lwid=c(0.2,1.1),
          scale = "none",margins = c(10,10),
          cexCol = 0.75,
          revC = TRUE,
          ColSideColors = my.heatcols,
          RowSideColors = t(my.heatcols[,3:1]),
          Rowv = my.hclust, Colv = my.hclust)


```

```{r}
# Date and time of Rendering
Sys.time()

sessionInfo()
```