---
title: "QC and filtering of spinal cord Lumbar 1"
author: "Fabio Sacher"
output:
  html_document:
    df_print: paged
  html_notebook:
    fig_height: 7
    fig_width: 8
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE, warning = FALSE, message = FALSE}
library(Seurat)
library(ggplot2)
library(dplyr)
library(scales)
library(gridExtra)
library(cowplot)
library(patchwork)
```

```{r input, message=FALSE, warning=FALSE, include=FALSE}
# This is where the raw data is located, results from a CellRanger pipeline
cellranger_path = "/scicore/home/tschoppp/GROUP/mapped_data/10x_NeuralTube_Lumbar_1_0420/"

#The sample name, will  be used in plots and files
my.samplename = "lumb1"
```

Date: `r format(Sys.Date(), "%d.%m.%y")`

This script takes raw UMIs from the 10x output, and runs some QC as well as filtering steps to remove cells which are likely doublets, very poor cells, as well as cells with hight abundance of MT read.

In the following plot we can see the distribution of UMIs per cell, in this raw UMI table count

```{r first look, echo=FALSE, message=FALSE, warning=FALSE}
ex = Read10X(data.dir = paste0(cellranger_path, "/outs/filtered_feature_bc_matrix/"), gene.column = 1)

ctable = ex

temp = data.frame(cells=colnames(ctable),UMIs=Matrix::colSums(ctable))
ggplot(temp, aes(x=temp$UMIs)) + geom_histogram() + geom_vline(aes(xintercept=mean(temp$UMIs))) +
  labs(title = paste0(my.samplename," Raw UMIs/Cell Frequency"), x = "UMIs/Cell", y = "Frequency")
rm(temp)
```

We first filter out the cells that have more than 4 times the mean of the UMI counts, and less than 20% the mean. These represent the probable doublets and poor cells.

The second filter is for mitochondrial UMIs, we get rid of cells that show that their fraction of UMIs from MT origin is higher than:   
The median + three times the MAD (median absolute deviation). 

The next filter is for cells that have an unusual number of genes detected per UMI

```{r cell filtering, message=FALSE, warning=FALSE, echo=FALSE}

my.plots=list()

summary(Matrix::colSums(ctable))

# Remove those cells that have more than X times the mean of UMI count
ctable = ctable[,Matrix::colSums(ctable)<(4*summary(Matrix::colSums(ctable))[4])]

#We remove the cells that have less than 20% of the reads median because they're likely false positives
ctable = ctable[,Matrix::colSums(ctable)>(0.2*summary(Matrix::colSums(ctable))[3])]

#Remove the cells that have more than 10% reads coming from mitochondrial genes
mt=c("ENSGALG00000035334", #COX3
     "ENSGALG00000032142", #MT-CO1
     "ENSGALG00000032079", #MT-CYB
     "ENSGALG00000037838", #ND6
     "ENSGALG00000029500", #ND5
     "ENSGALG00000036229", #MT-ND4
     "ENSGALG00000042478", #ND4L
     "ENSGALG00000030436", #ND3
     "ENSGALG00000041091", #MT-ATP6
     "ENSGALG00000032465", #MT-ATP8
     "ENSGALG00000043768", #MT-ND2
     "ENSGALG00000042750") #MT-ND1

is.mt = (rownames(ctable)%in%mt)
percent.mt = Matrix::colSums(ctable[is.mt, ])/Matrix::colSums(ctable)

my.threshold = median(percent.mt) + (mad(percent.mt) * 3)

my.filter1=rep(1,length(percent.mt))
my.filter1[which(percent.mt> my.threshold & Matrix::colSums(ctable) < median(Matrix::colSums(ctable)) )] = 0
my.filter1 =as.factor(my.filter1)

toplot = data.frame(mt = percent.mt, UMI = Matrix::colSums(ctable))
my.scale = max(table(cut(toplot$mt, breaks = round(max(toplot$mt)/0.005))))

my.plots[[1]] = ggplot() +
  geom_point(mapping = aes(color=my.filter1, x=toplot$mt, y=(toplot$UMI*my.scale)/max(toplot$UMI))) +
  geom_freqpoly(mapping = aes(x=toplot$mt),binwidth=0.005, color ="blue") +
  annotate("text", x = my.threshold, y = 0, label = round(my.threshold,2), color="red2") +
  scale_y_continuous(name = "UMI counts", labels = function(b) { round(b*max(toplot$UMI)/my.scale)},
                     sec.axis = sec_axis(~.,name = "Cell frequency")) +
  scale_color_manual(values=c("red2", "gray20")) +
  labs(title = paste0(my.samplename," MT% vs UMIs relation"), x = "MT%") +
  theme(legend.position="none", axis.title.y.right = element_text(color = "blue")) 

ctable=ctable[,-which(percent.mt> my.threshold & Matrix::colSums(ctable) < median(Matrix::colSums(ctable)) )]

gct = data.frame(cell = colnames(ctable), genes = (0))

# Put the gene counts
for (i in 1:dim(ctable)[2] ){
  gct [i,2] = length(which(ctable[,i] > 0))
}

#Column for UMI counts
gct = cbind(gct, UMIs=(Matrix::colSums(ctable)))

my.filter=rep("black",dim(gct)[1])
my.filter[which((gct$genes/gct$UMIs)<0.15 & gct$genes<(max(gct$genes)/3)*2)] = "red2"

my.plots[[2]] = ggplot(gct, aes(x=UMIs, y=genes)) + geom_point(color=my.filter) + geom_smooth() +
  geom_smooth(method = "lm", formula = y~x, color="red") + expand_limits(y=0, x=0) +
  labs(title = paste0(my.samplename," Genes vs UMIs relation"), x = "UMIs", y = "Genes") +
  geom_abline(intercept = 0, slope = 0.15, color="red", linetype="dashed", size=0.5) + theme_bw()

grid.arrange(grobs=my.plots, ncol=2)

#Remove cells for which the UMI counts are not substancially adding to the gene count (many counts on very few genes)
ctable = ctable[,-which((gct$genes/gct$UMIs)<0.15 & gct$genes<(max(gct$genes)/3)*2)]

gct=gct[gct$cell %in% colnames(ctable),]

```

 In this case the mitochondrial percent cutoff is `r round(median(percent.mt) + (mad(percent.mt) * 3),3)` .

After all these filtering steps, we can see the distribution of UMI counts, the UMI/gene frequency, the distribution of genes detected per cell and the final UMI/gene relation

```{r final plots, message=FALSE, warning=FALSE, echo=FALSE}
temp1 = data.frame(cells=colnames(ctable),UMIs=Matrix::colSums(ctable))
my.plots[[1]] = ggplot(temp1, aes(x=temp1$UMIs)) + geom_histogram() + geom_vline(aes(xintercept=mean(temp1$UMIs))) +
  labs(title = paste0(my.samplename," Filtered UMIs/Cell Frequency"), x = "UMIs/Cell", y = "Frequency") + theme_bw()

temp = data.frame(genes=rownames(ctable), UMIs=log(Matrix::rowMeans(ctable)))
my.plots[[2]] = ggplot(temp, aes(x=temp$UMIs)) + geom_histogram() +
  labs(title = paste0(my.samplename," Filtered UMIs/Gene (log) Frequency"), x = "UMIs/Gene", y = "Frequency") + theme_bw()

# Distribution of genes
my.plots[[3]] = ggplot(gct, aes(x=genes)) + 
  geom_histogram(aes(y=(..count..)/sum(..count..))) + 
  scale_y_continuous(labels = percent) +
  geom_vline(aes(xintercept=mean(genes))) + 
  labs(title = paste0(my.samplename," Genes/Cell frequency"), x = "Number of genes", y ="Frequency") + theme_bw()

# Gene count / UMI count, with a linear regression
my.plots[[4]] = ggplot(gct, aes(x=UMIs, y=genes)) + geom_point() + geom_smooth() +
  geom_smooth(method = "lm", formula = y~x, aes(color="red")) + expand_limits(y=0, x=0) +
  labs(title = paste0(my.samplename," Genes vs UMIs relation"), x = "UMIs", y = "Genes") + theme_bw() + theme(legend.position="none")

grid.arrange(grobs=my.plots, ncol=2)
```

### Save the dataset
We will save the dataset as data.rds, which can then be imported into the scRNAseq pipeline of your choice for downstream analysis.
```{r save}
saveRDS(data, file = paste0("~/spinal_cord_paper/data/",my.samplename,"_filtered_",format(Sys.Date(), "%d%m%y"),".rds"))
```

```{r sessionInfo}
# Date and time of Rendering
Sys.time()

sessionInfo()
```

